### DISCUSSION TITLE
Ways to detect mistakes in TTS outputs?

### DISCUSSION THREAD

**Discussion by erogol** [coqui-ai/TTS#801] [General Q&A]
I wonder if there is any literature on detecting inconsistencies in TTS outputs to monitor & secure the deployed system performance.

Feel free to share ideas and experiences too so that we can dabble on together.

**Comment by talipturkmen** [comment]
I think you can use monitor acoustic estimator metrics like snr.

**Reply by erogol** [reply]
Good idea... You think it would also help for detecting prosody problems?

**Reply by talipturkmen** [reply]
For prosody, I'm not so sure since it's more related to durations and the pitch. But it can help to reduce noise artifacts.

**Reply by Sadam1195** [reply]
> Good idea... You think it would also help for detecting prosody problems?

As @talipturkmen pointed out for prosody related problems you can take the following features for consideration as they are more related to energy and audio patterns hidden also vaguely discussed in the paper here https://aclanthology.org/L16-1211.pdf
 
https://librosa.org/doc/main/generated/librosa.feature.rms.html#librosa-feature-rms
https://librosa.org/doc/main/generated/librosa.feature.spectral_centroid.html#librosa-feature-spectral-centroid
https://librosa.org/doc/main/generated/librosa.feature.chroma_cens.html#librosa-feature-chroma-cens

**Reply by erogol** [reply]
Thanks for the paper. I'll read and get back

**Reply by erogol** [reply]
An interesting approach could be training a one-class SVM to detect anomaly model outputs as the model is functioning using the features you pointed above. But this approach assumes the model works generally fine and there may be some mistakes here and there.

**Reply by Sadam1195** [reply]
> But this approach assumes the model works generally fine and there may be some mistakes here and there.

Perhaps it's worth a try!

**Reply by Sadam1195** [reply]
> SVM

I would prefer Gaussian Mixer Models (GMM) for that. When it comes to sound/speech (in my research and experiments) I have seen GMMs out-perform many sophisticated models like SVMs etc

**Reply by Sadam1195** [reply]
> An interesting approach could be training a one-class SVM to detect anomaly model outputs as the model is functioning using the features you pointed above.

or one thing that could be tried is to measure those features and take a ratio between train and eval audios and keep a dict of values to monitor the rate of change to see if the change is monotonically increasing/decreasing and see if that could be used for determining the criteria to evaluate the prosody.

**Reply by erogol** [reply]
But these values are not text invariant right? So diff sentences can change the values dramatically. Then how we can compare the eval results with train results?

**Reply by Sadam1195** [reply]
> But these values are not text invariant right? 

No. texts are irrelevant when evaluating the quality of single instance of speech. We are only interested in audio features of the instance. 
> So diff sentences can change the values dramatically. Then how we can compare the eval results with train results?

Yes, but ratios would be consistent as I am considering only audio part of the outputs.
They are two parts of this problem. (A ; train instances, B ; eval instances). Consider this, say model is in training you could determine if the quality is improving of  _(A ; train instances)_ with epochs increasing given in the light of features discussed above. So if the ratio of A part with B part is increasing/decreasing monotonically you can determine the correlation since the B part is kind of constant. You could remove the B part from equation but I am assuming it would play important role as it is a reference point.

**Reply by erogol** [reply]
I am still confused about how we use features to measure the quality. I guess to do that we need a reference to compare. Then where is that reference? 

If we use something similar to the paper above, they train a classifier over a set of features to measure the TTS quality. But then where is the data we train these models initially? Is it the training data? 

Sorry for the confusion.... It also has been a while. I might have forgotten a token or two about our discussion :)

**Reply by Sadam1195** [reply]
Can we discuss this over a call, would you prefer that?

**Reply by erogol** [reply]
Sure. Check gitter. I've sent you a link.
